torchvision/datasets/sun397.py

```diff
@@ -0,0 +1,108 @@
+
+    Args:
+        root (string): Root directory of the dataset.
+        split (string, optional): The dataset split, supports ``"train"`` (default) and ``"test"``.
+        parition (string, integer, optional): A valid partition can be an integer from 1 to 10 or ``"all"``
+            for the entire dataset.
+        transform (callable, optional): A function/transform that  takes in an PIL image and returns a transformed
+            version. E.g, ``transforms.RandomCrop``.
+        target_transform (callable, optional): A function/transform that takes in the target and transforms it.
+    """
+
+    _URL = "http://vision.princeton.edu/projects/2010/SUN/SUN397.tar.gz"
+    _FILENAME = "SUN397.tar.gz"
+    _MD5 = "8ca2778205c41d23104230ba66911c7a"
+    _PARTITIONS_URL = "https://vision.princeton.edu/projects/2010/SUN/download/Partitions.zip"
+    _PARTITIONS_FILENAME = "Partitions.zip"
+
+    def __init__(
+        self,
+        root: str,
+        split: str = "train",
+        partition: Optional[int] = 1,
+        download: bool = True,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(root, transform=transform, target_transform=target_transform)
+        self.split = verify_str_arg(split, "split", ("train", "test"))
+        self.partition = partition
+        self.data_dir = Path(self.root) / "SUN397"
+
+        if self.partition is not None:
+            if self.partition < 0 or self.partition > 10:
+                raise RuntimeError("Enter a valid integer partition from 1 to 10 or None, for entire dataset")
+
+        if download:
+            self._download()
+
+        if not self._check_exists():
+            raise RuntimeError("Dataset not found. You can use download=True to download it")
+
+        with open(self.data_dir / "ClassName.txt", "r") as f:
+            self.classes = [c[3:].strip() for c in f]
+
+        self.class_to_idx = dict(zip(self.classes, range(len(self.classes))))
+
+        self._labels = []
+        self._image_files = []
+        if self.partition is not None:
+            with open(self.data_dir / f"{self.split.title()}ing_{self.partition:02d}.txt", "r") as f:
+                self._image_files, self._labels = zip(
+                    *(
+                        (
+                            self.data_dir.joinpath(*posix_rel_path.split("/")),
+                            self.class_to_idx["/".join(Path(posix_rel_path).parts[1:-1])],
+                        )
+                        for posix_rel_path in (line.strip()[1:] for line in f)
+                    )
+                )
+
+        else:
+            for path, _, files in os.walk(self.data_dir):
+                for file in files:
+                    if file[:3] == "sun":
+                        self._image_files.append(Path(path) / file)
+                        self._labels.append(Path(path).relative_to(self.data_dir).as_posix()[2:])
```

Comment:
We can at least oust the label generation

```suggestion
                if self.partition is not None:
            with open(self.data_dir / f"{self.split.title()}ing_{self.partition:02d}.txt", "r") as f:
                self._image_files = [self.data_dir.joinpath(*line.strip()[1:].split("/")) for line in f]
        else:
            self._image_files = list(self.data_dir.rglob("sun_*.jpg"))
        self._labels = [
            self.class_to_idx["/".join(path.relative_to(self.data_dir).parts[1:-1])] for path in self._image_files
        ]
```